File: FAB/FAB.Model.FSharp/MissileFunctions.fs
module MissileFunctions
open FAB.Model

let generateBombSpread centreCol centreRow board =
    let colRange = [centreCol - 1..centreCol + 1]
    let rowRange = [centreRow - 1..centreRow + 1]
    [for col in colRange do
          for row in rowRange do
            yield new Location(col, row ) ]

let fireBomb (loc: Location) board = 
    let locs = generateBombSpread loc.Col loc.Row board
    GameBoardFunctions.checkSquaresAndRecordOutcome board locs

let fireMissile (loc: Location) board =
    GameBoardFunctions.checkSquareAndRecordOutcome board loc false

File: FAB/FAB.Model.CSharp/MissileFunctions.cs
using Quadrivia.FunctionalLibrary;
using System.Collections.Immutable;
using System.Linq;

namespace FAB.Model
{
    public class MissileFunctions
    {
        public static GameBoard fireMissile(Location loc, GameBoard board)
        {
            return board.checkSquareAndRecordOutcome(loc);
        }

        public static GameBoard fireBomb(Location loc, GameBoard board)
        {
            return board.checkSquaresAndRecordOutcome(GenerateLocationsToHit(loc.Col, loc.Row, board));
        }

        private static FList<Location> GenerateLocationsToHit(int centreCol, int centreRow, GameBoard board)
        {
            return FList.New(Enumerable.Range(centreCol - 1, 3)
                .SelectMany(col => Enumerable.Range(centreRow - 1, 3), 
                (col, row) => new Location(col, row))
                .ToArray());
        }
    }
}


